include/exadg/functions_and_boundary_conditions/function_cached.cpp

```diff
@@ -24,53 +24,85 @@
+                                            unsigned int const face,
+                                            unsigned int const q,
+                                            unsigned int const v) const
 {
-  Assert(map_map_vector_index != nullptr,
-         dealii::ExcMessage("Pointer global_map_vector_index is not initialized."));
-  Assert(map_map_vector_index->find(quad_index) != map_map_vector_index->end(),
-         dealii::ExcMessage("Specified quad_index does not exist in global_map_vector_index."));
+  Assert(map_vector_index.find(q_index) != map_vector_index.end(),
+         dealii::ExcMessage("Specified q_index does not exist in map_vector_index."));
 
-  Assert(map_array_solution != nullptr,
-         dealii::ExcMessage("Pointer map_array_solution is not initialized."));
-  Assert(map_array_solution->find(quad_index) != map_array_solution->end(),
-         dealii::ExcMessage("Specified quad_index does not exist in map_array_solution."));
-
-  MapVectorIndex const &      map_vector_index = map_map_vector_index->find(quad_index)->second;
-  ArraySolutionValues const & array_solution   = map_array_solution->find(quad_index)->second;
+  Assert(map_solution.find(q_index) != map_solution.end(),
+         dealii::ExcMessage("Specified q_index does not exist in map_solution."));
 
   Id                              id    = std::make_tuple(face, q, v);
-  dealii::types::global_dof_index index = map_vector_index.find(id)->second;
+  dealii::types::global_dof_index index = map_vector_index.find(q_index)->second.find(id)->second;
 
+  ArraySolutionValues const & array_solution = map_solution.find(q_index)->second;
   Assert(index < array_solution.size(), dealii::ExcMessage("Index exceeds dimensions of vector."));
 
   return array_solution[index];
 }
 
-template<int rank, int dim, typename Number>
+template<int rank, int dim>
+std::vector<typename ContainerInterfaceData<rank, dim>::quad_index> const &
+ContainerInterfaceData<rank, dim>::get_quad_indices()
+{
+  return quad_indices;
+}
+
+template<int rank, int dim>
+typename ContainerInterfaceData<rank, dim>::ArrayQuadraturePoints &
+ContainerInterfaceData<rank, dim>::get_array_q_points(quad_index const & q_index)
+{
+  return map_q_points[q_index];
+}
+
+template<int rank, int dim>
+typename ContainerInterfaceData<rank, dim>::ArraySolutionValues &
+ContainerInterfaceData<rank, dim>::get_array_solution(quad_index const & q_index)
+{
+  return map_solution[q_index];
+}
+
+template<int rank, int dim>
+FunctionCached<rank, dim>::FunctionCached()
+{
+}
+
+template<int rank, int dim>
+typename FunctionCached<rank, dim>::value_type
+FunctionCached<rank, dim>::tensor_value(unsigned int const face,
+                                        unsigned int const q,
+                                        unsigned int const v,
+                                        unsigned int const quad_index) const
+{
+  return interface_data->get_data(quad_index, face, q, v);
+}
```

Comment:
I guess you want to move this function into the header file and inline it.

Reply:
you mean for performance reasons? why is it essential to have it in the header file?

Reply:
see https://github.com/exadg/exadg/pull/230/commits/99c3b18907d85d234e43e987e77ecb606abb728f